plugins/core/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/FeatureDevSessionContext.kt

```diff
@@ -49,40 +59,75 @@ class FeatureDevSessionContext(val project: Project) {
     private val gitIgnoreFile = File(projectRoot.path, ".gitignore")
 
     init {
-        ignorePatternsWithGitIgnore = ignorePatterns + parseGitIgnore().map { Regex(it) }
+        ignorePatternsWithGitIgnore = (ignorePatterns + parseGitIgnore().map { Regex(it) }).toList()
     }
 
     fun getProjectZip(): ZipCreationResult {
-        val zippedProject = runReadAction { zipFiles(projectRoot) }
+        val zippedProject = runBlocking { zipFiles(projectRoot) }
         val checkSum256: String = Base64.getEncoder().encodeToString(DigestUtils.sha256(FileInputStream(zippedProject)))
         return ZipCreationResult(zippedProject, checkSum256, zippedProject.length())
     }
 
-    fun ignoreFile(file: File): Boolean = try {
-        ignorePatternsWithGitIgnore.any { p -> p.containsMatchIn(file.path) }
-    } catch (e: Exception) {
-        true
+    private suspend fun ignoreFile(file: File): Boolean = coroutineScope {
+        val deferredResults = ignorePatternsWithGitIgnore.map { pattern ->
+            async {
+                pattern.containsMatchIn(file.path)
+            }
+        }
+        deferredResults.any { it.await() }
     }
 
-    fun ignoreFile(file: VirtualFile): Boolean = ignoreFile(File(file.path))
+    suspend fun ignoreFile(file: VirtualFile): Boolean = ignoreFile(File(file.path))
+
+    suspend fun zipFiles(projectRoot: VirtualFile): File = withContext(Dispatchers.Default) {
+        val files = mutableListOf<VirtualFile>()
+        VfsUtil.visitChildrenRecursively(projectRoot, object : VirtualFileVisitor<Void?>() {
+            override fun visitFile(file: VirtualFile): Boolean {
+                if (file.isFile) {
+                    files.add(file)
+                    return true
+                }
+                return runBlocking {
+                    !ignoreFile(file)
+                }
+            }
+        })
+
+        // Process files in parallel
+        val nonMatchingFiles = mutableListOf<VirtualFile>()
+        coroutineScope {
+            files.map { file ->
+                async {
+                    if (file.isFile && !ignoreFile(file)) {
+                        nonMatchingFiles.add(file)
+                    }
+                }
+            }.awaitAll()
```

Comment:
don't see the point of the coroutine if we end up waiting for everything anyways
use a sequence

https://kotlinlang.org/docs/sequences.html#from-chunks

Reply:
does yield also execute ignoreFile(file) in parallel?

Reply:
i think you want
```
channelFlow {
    files.chunked(50).forEach { // chunk with some reasonable number because we don't actually need a new job for each file
        launch {
            for(file in files) {
                if (file.isFile && !ignoreFile(file)) {
                    send(file)
                }
            }
        }
    }
}
```

Reply:
```
val flow = channelFlow {
            // chunk with some reasonable number because we don't actually need a new job for each file
            files.chunked(50).forEach { chunk ->
                launch {
                    for(file in chunk) {
                        if (file.isFile && !ignoreFile(file, this)) {
                            send(file)
                        }
                    }
                }
            }
        }
        flow.collect {
            filesToInclude.add(it)
        }
```
this?

Reply:
more like:

```kt
val filesToIncludeFlow = channelFlow {
...
}

createTemporaryZipFileAsync {
    filesToIncludeFlow.collect { file ->
    ...
    }
}
```

Reply:
There's a specific scenario that when processing files with really long paths, the processing time would also increase significantly for each file. And the processing for those will be grouped together because they are siblings.

Reply:
idea is you avoid a terminal operation like collect until the point you actually need the result of the pipeline

Reply:
It would be ideal if executing of those would be parallelized as much as possible. That seems to be the bottleneck I'm seeing when processing large projects.

Reply:
let's do a follow up later then because it looks like what's happening is:
```
START                                             END
|                                                 |
|----- collect files ------|
                           |------- make zip -----| 
```

but what we should be doing is building and streaming the zip upload as we traverse the project and find files that should be included

Reply:
sure